### PR TITLE
Ensure that the IsingXX and IsingZZ gates are differentiable

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**
The gates were recently added `qml.IsingXX `in #1194 ,`qml.IsingZZ `in #1199 but there are at the moment not differentiable. This PR is directly related to the following issue https://github.com/PennyLaneAI/pennylane/issues/1386 .

**Description of the Change:**
1) I added functions representing these operations to `jax_ops.py,` `autograd_ops.py`, and `tf_ops.py`.
2) I modified the `DefaultQubitAutograd`, `DefaultQubitJAX`, and `DefaultQubitTF `devices to support these differentiable operations.
3) [Wip] Tests that ensure that the gradient of IsingXX and IsingZZ are correct for `diff_method="parameter-shift"` and `diff_method="backprop"`, for all ML interfaces (autograd, TF, JAX).

**Related issues**

Closes https://github.com/PennyLaneAI/pennylane/issues/1386.

### CODE REVIEW

**Review by antalszava — COMMENTED** [review]
Hi @rmoyard, this is overall amazing! :tada: :blush: Very nice to have this support and the testing using analytic expressions for the gradients is solid. :rocket: 

Listed a couple of thoughts for improvement, more important ones seemed to be:
* moving the tests to the dedicated operation test file and
* double-checking if adjoint differentiation could also be supported.

Also, don't forget to update the `.github/CHANGELOG.md` file by detailing the addition (it could go into the Bug fixes section for example) and adding yourself as a contributor. :slightly_smiling_face:

File: tests/interfaces/test_qnode_jax.py [hunk 1]
```diff
@@ -87,6 +87,78 @@ def circuit(weights):
     assert "DeviceArray" in val.__repr__()
 
 
+@pytest.mark.parametrize(
+    "dev_name,diff_method",
+    [
+        ["default.qubit", "backprop"],
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/interfaces/test_qnode_jax.py:93]: The use of `["default.qubit", "parameter-shift"] ` crashes Python, Josh told me that maybe @albi3ro could help on this one?

File: tests/interfaces/test_qnode_autograd.py [hunk 2]
```diff
@@ -1073,6 +1073,74 @@ def circuit(x):
         jac_fn(x)
         assert j_spy.call_count == 5
 
+    def test_grad_ising_xx(self, dev_name, diff_method, tol):
+        """Test the gradient for the gate IsingXX."""
+        if diff_method in {"adjoint"}:
+            pytest.skip("Test does not support adjoint")
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/interfaces/test_qnode_autograd.py:1079]: Yes exactly the problem is that operations do not have generators, e.g. `ValueError: Operation IsingZZ does not have a generator`. I will check if I can add them.

File: tests/interfaces/test_qnode_autograd.py [hunk 3]
```diff
@@ -1073,6 +1073,74 @@ def circuit(x):
         jac_fn(x)
         assert j_spy.call_count == 5
 
+    def test_grad_ising_xx(self, dev_name, diff_method, tol):
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/interfaces/test_qnode_autograd.py:1076]: Ok I'll move all the tests in this folder!

File: tests/interfaces/test_qnode_autograd.py [hunk 4]
```diff
@@ -1073,6 +1073,74 @@ def circuit(x):
...
+        norm = np.linalg.norm(init_state)
+        init_state /= norm
+
+        @qml.qnode(dev, diff_method=diff_method, interface="autograd")
+        def circuit(phi):
+            qml.QubitStateVector(init_state, wires=[0, 1])
+            qml.IsingXX(phi, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        phi = np.array(0.1, requires_grad=True)
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/interfaces/test_qnode_autograd.py:1098]: I added the option for phi.

File: tests/interfaces/test_qnode_autograd.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/interfaces/test_qnode_autograd.py]: I added both generators and it is working 👍

File: pennylane/ops/qubit.py [hunk 5]
```diff
@@ -1797,6 +1797,7 @@ class IsingXX(Operation):
     num_wires = 2
     par_domain = "R"
     grad_method = "A"
+    generator = [np.array([[0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0]]), -1 / 2]
```
Comment (antalszava) [inline pennylane/ops/qubit.py:1800]: :partying_face:

File: tests/ops/test_qubit_ops.py [hunk 6]
```diff
@@ -777,6 +778,241 @@ def test_isingxx_decomposition(self, tol):
...
+
+        phi = npp.array(0.1, requires_grad=True)
+
+        expected = (1 / norm ** 2) * (-2 * (psi_0 * psi_2 + psi_1 * psi_3) * np.sin(phi))
+
+        res = qml.grad(circuit)(phi)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("dev_name,diff_method,phi", configuration)
+    def test_isingxx_jax_(self, tol, dev_name, diff_method, phi):
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:861]:
```suggestion
    def test_isingxx_jax_grad(self, tol, dev_name, diff_method, phi):
```

File: tests/ops/test_qubit_ops.py [hunk 7]
```diff
@@ -777,6 +778,241 @@ def test_isingxx_decomposition(self, tol):
...
+        """Test the gradient for the gate IsingXX."""
+
+        if diff_method in {"finite-diff"}:
+            pytest.skip("Test does not support parameter-shift")
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:865]: The message seems to be off here. Would this be the same issue as for parameter-shift when using `jax`?
```suggestion
            pytest.skip("Test does not support finite-diff")
```

File: tests/ops/test_qubit_ops.py [hunk 8]
```diff
@@ -777,6 +778,241 @@ def test_isingxx_decomposition(self, tol):
...
+
+        if diff_method in {"parameter-shift"}:
+            pytest.skip("Test does not support parameter-shift")
+
+        jax = pytest.importorskip("jax")
+        jnp = pytest.importorskip("jax.numpy")
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:871]: Nice!

File: tests/ops/test_qubit_ops.py [hunk 9]
```diff
@@ -777,6 +778,241 @@ def test_isingxx_decomposition(self, tol):
...
+
+        res = jax.grad(circuit, argnums=0)(phi)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("dev_name,diff_method,phi", configuration)
+    def test_isingzz_jax_grad(self, tol, dev_name, diff_method, phi):
+        """Test the gradient for the gate IsingZZ."""
+
+        if diff_method in {"finite-diff"}:
+            pytest.skip("Test does not support parameter-shift")
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:912]:
```suggestion
            pytest.skip("Test does not support finite-diff")
```

**Review by antalszava — COMMENTED** [review]
Looks great @rmoyard! :slightly_smiling_face:  Might check back for the `jax` parameter-shift update before approval, but otherwise no blockers from my side.

File: tests/interfaces/test_qnode_jax.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (josh146) [reply tests/interfaces/test_qnode_jax.py]: Yes, this is a very concerning bug! @rmoyard, does Python still crash on the GitHub CI, or just locally?

If we can’t resolve it, what we could do is mark the JAX+parameter-shift case as `xfail`, and open an issue. That way we can still merge this PR in, as it contains an important bugfix for the upcoming release.

[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/interfaces/test_qnode_jax.py]: Apparently the bug comes from the fact that Jax tries to train the initial state `init_state = jnp.array([psi_0, psi_1, psi_2, psi_3])` and you don't have a parameter to specify that it is not the case like in Pennylane numpy `requires_grad=False`.

**Review by antalszava — APPROVED** [review]
Looks good to me! :+1: :partying_face: Thank you for getting this through. For Jax, we could create an issue and return to it at a later stage.

**Pull request merged**